### PR TITLE
Urlmageddon

### PR DESCRIPTION
**Pull request by emilio**
Still needs a bunch of code in net to be converted in order to get more
advantage of this for images and stuff, but meanwhile this should help quite a
bit with #13778.

Still wanted to get this in.

r? @SimonSapin

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14246)

### CODE REVIEW

File: components/style_traits/lib.rs [hunk 1]
```diff
@@ -18,12 +18,44 @@ extern crate app_units;
...
+///
+/// FIXME: Better name for this appreciated.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf, Serialize, Deserialize))]
+pub enum ServoUrl {
+    Data(Arc<String>),
+    Url(Arc<Url>),
+}
+
+impl From<Url> for ServoUrl {
```
Comment (emilio) [inline components/style_traits/lib.rs]: This is just so transition is easier, and needs to be removed afterwards.

**Comment by SimonSapin** [comment]
While you’re at it, is replacing every use of `url::Url` at once with sed much harder? `ServoUrl` could be made to have some APIs compatible with `url::Url` (like `parse`, `join`, …) to ease the transition.

---

Reviewed 17 of 20 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[components/net/image_cache_thread.rs, line 540 at r2](https://reviewable.io:443/reviews/servo/servo/14246#-KWhS-Rfi5GIkZcCdJe5:-KWhS-Rfi5GIkZcCdJe6:bq6sjcb) ([raw file](https://github.com/servo/servo/blob/be6242081b24ad51350ac253d250f441bf17edb1/components/net/image_cache_thread.rs#L540)):*
> ```Rust
>                         // TODO(emilio): ServoUrl in more places please!
>                         let request = RequestInit {
>                             url: Url::parse(url.as_str()).unwrap(),
> ```

This (and a couple lines below) needlessly goes through rust-url’s parser even when the `ServoUrl` value is a non-data URL. This should instead use a conversion method on `ServoUrl`, even if that method is removed eventually.

---

*[components/style/values/specified/url.rs, line 91 at r2](https://reviewable.io:443/reviews/servo/servo/14246#-KWhQgNDi-pKuPEY-ciS:-KWhQgNEjhWwYBIUBBJl:bi6fecl) ([raw file](https://github.com/servo/servo/blob/be6242081b24ad51350ac253d250f441bf17edb1/components/style/values/specified/url.rs#L91)):*
> ```Rust
> 
>         let serialization = Arc::new(url.into_owned());
>         let resolved = if serialization.starts_with("data:") {
> ```

This `if` / `else` block should be in a method or constructor of `ServoUrl`. Perhaps mimicking `url::Url`’s API: a `parse` constructor that doesn’t use a base URL, and a `join` method that does.

---

*[components/style_traits/Cargo.toml, line 21 at r2](https://reviewable.io:443/reviews/servo/servo/14246#-KWhR1q1uof17sy-3Sl4:-KWhR1q2AF4ymOli_dqR:b-3s4hgr) ([raw file](https://github.com/servo/servo/blob/be6242081b24ad51350ac253d250f441bf17edb1/components/style_traits/Cargo.toml#L21)):*
> ```TOML
> cssparser = "0.7"
> euclid = "0.10.1"
> url = "1.2"
> ```

This likely needs a `./mach cargo-update -p style_traits` run to update all lock files.

---

*[components/style_traits/lib.rs, line 36 at r2](https://reviewable.io:443/reviews/servo/servo/14246#-KWhQ5F9FZcFiyXTPu2V:-KWhQ5F9FZcFiyXTPu2W:b-eoazoo) ([raw file](https://github.com/servo/servo/blob/be6242081b24ad51350ac253d250f441bf17edb1/components/style_traits/lib.rs#L36)):*
> ```Rust
> /// afterwards.
> ///
> /// FIXME: Better name for this appreciated.
> ```

I think it’s OK to call this `Url`. Types are namespaced in modules and crates, let’s take advantage of that. (Only a few modules will need to rename one of the two types in order to import both.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14246#-:-KWhTbM-fhszp-ruL0SK:bmcdofa)*

**Comment by emilio** [comment]
@bors-servo try

I was really dubious on the data thing after seeing some use cases, so I decided to defer it for after the landing on this with more discussion, because I think, for example, we should still percent-decode, probably when serializing too. So I'd rather land this (a wrapper for an `Arc<Url>` with the proper interface we need (and indeed I was using until I found that out), and then do the switch to `enum { Data(..), Url(..) }`.

But I think this is mostly ready for review (given tests pass), and I'd appreciate a fast review given how prone it is to bitrot.

**Comment by emilio** [comment]
@bors-servo retry

 * Forgot to add some files.

**Comment by emilio** [comment]
@bors-servo try

 * OSX-only bustage

**Comment by emilio** [comment]
@bors-servo try

 * OSX

**Comment by emilio** [comment]
@bors-servo try 

 * CEF, of course.

**Comment by SimonSapin** [comment]
Reviewed 160 of 161 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[components/gfx/font_cache_thread.rs, line 211 at r3](https://reviewable.io:443/reviews/servo/servo/14246#-KWmo1-mgQADtlYtRKLF:-KWmo1-mgQADtlYtRKLG:b-jtru2q) ([raw file](https://github.com/servo/servo/blob/536a264bd89a2e5ac2be8b85801b6917bc2631f4/components/gfx/font_cache_thread.rs#L211)):*
> ```Rust
>                 let url = match url_source.url.url() {
>                     Some(url) => url.clone(),
>                     None => return,
> ```

Please add a comment that says this happens when URL parsing fails.

---

*[components/layout_thread/Cargo.toml, line 27 at r3](https://reviewable.io:443/reviews/servo/servo/14246#-KWmld46Dzty-V293eAR:-KWmld47Rt-VQWeh2a_q:b1af9yl) ([raw file](https://github.com/servo/servo/blob/536a264bd89a2e5ac2be8b85801b6917bc2631f4/components/layout_thread/Cargo.toml#L27)):*
> ```TOML
> msg = {path = "../msg"}
> net_traits = {path = "../net_traits"}
> style_traits = {path = "../style_traits"}
> ```

I think this is not needed anymore.

---

*[components/script/script_thread.rs, line 1756 at r3](https://reviewable.io:443/reviews/servo/servo/14246#-KWmrfdbBvIRhEmJ435o:-KWmrfdcD-TsfhWOH3Hu:b2usozl) ([raw file](https://github.com/servo/servo/blob/536a264bd89a2e5ac2be8b85801b6917bc2631f4/components/script/script_thread.rs#L1756)):*
> ```Rust
>             // append question mark and query component, if any;
>             // append number sign and fragment component if any.
>             let encoded = &incomplete.url.as_url().unwrap()[Position::BeforePath..];
> ```

Should this really unwrap?

---

*[components/script/dom/response.rs, line 361 at r3](https://reviewable.io:443/reviews/servo/servo/14246#-KWmsJhFaznt1cKUxm8q:-KWmsJhFaznt1cKUxm8r:b2usozl) ([raw file](https://github.com/servo/servo/blob/536a264bd89a2e5ac2be8b85801b6917bc2631f4/components/script/dom/response.rs#L361)):*
> ```Rust
> 
> fn serialize_without_fragment(url: &ServoUrl) -> &str {
>     &url.as_url().unwrap()[..Position::AfterQuery]
> ```

Should this really unwrap?

---

*[components/style/values/specified/url.rs, line 125 at r3](https://reviewable.io:443/reviews/servo/servo/14246#-KWmtCR_Q_nDp-K90ZyG:-KWmtCR_Q_nDp-K90ZyH:b-wqfwbs) ([raw file](https://github.com/servo/servo/blob/f14e7339b5ff95fce0127dce4fe87ce082ab7259/components/style/values/specified/url.rs#L125)):*
> ```Rust
>     }
> 
>     // Just for unit tests, don't use outside of them!
> ```

Why can this comment be removed now? (And why was it there in the first place?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14246)*

**Comment by SimonSapin** [comment]
@bors-servo r+ p=1000

---

Reviewed 1 of 161 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14246#-:-KWmuMKCcg-oJ3TUS47Q:bdnb5fs)*

**Comment by emilio** [comment]
@bors-servo r=SimonSapin

 * ARRRGGHH, test-stylo doesn't run with `./mach test-unit` :((

**Comment by KiChjang** [comment]
@bors-servo retry
* #10473

**Pull request merged**